File: _deploy/scripts/scraping/scrape-canlii-tribunals.js
#!/usr/bin/env node
/**
 * CanLII Tribunal Decision Scraper (Node.js version)
 * Scrapes WSIAT, HRTO, SST decisions to seed the 3 Flywheels
 * 
 * FREE TOOLS ONLY - NO API COSTS
 * Author: 3mpwrApp
 * Date: April 2026
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// ===== CONFIGURATION =====
const CANLII_BASE = "https://api.canlii.org/v1";
const CANLII_API_KEY = process.env.CANLII_API_KEY || "YOUR_FREE_API_KEY_HERE";
const OUTPUT_DIR = path.join(__dirname, "../data/tribunal-decisions");
const BATCH_SIZE = 50;

// MAXIMUM COLLECTION MODE: Get EVERYTHING from ALL of Canada
// User wants: "all data from all provinces covering all of canada from the oldest year that they have to current NO EXCEPTIONS"
// Changed from "2000-01-01" to "1900-01-01" to capture OLDEST available decisions
const CHANGED_SINCE = "1900-01-01"; // Some tribunals have decisions dating back to early 1900s

const TRIBUNALS = {
  // === ONTARIO === (ALREADY COLLECTED - SKIPPING)
  // "onwsiat": { ... },
  // "onhrt": { ... },
  // "onca": { ... },
  
  // === BRITISH COLUMBIA ===
  "bchrt": {
    "name": "British Columbia Human Rights Tribunal",
    "database": "bchrt",
    "jurisdiction": "BC",
    "search_terms": ["accommodation", "disability"]
  },
  "bcwcat": {
    "name": "Workers' Compensation Appeal Tribunal (BC)",
    "database": "bcwcat",
    "jurisdiction": "BC",
    "search_terms": ["chronic pain", "PTSD", "back injury", "disability"]
  },
  "bcca": {
    "name": "British Columbia Court of Appeal",
    "database": "bcca",
    "jurisdiction": "BC",
    "search_terms": ["disability", "WorkSafeBC"]
  },
  
  // === ALBERTA ===
  "abqb": {
    "name": "Alberta Court of Queen's Bench",
    "database": "abqb",
    "jurisdiction": "AB",
    "search_terms": ["disability", "WCB", "accommodation"]
  },
  "abca": {
    "name": "Alberta Court of Appeal",
    "database": "abca",
    "jurisdiction": "AB",
    "search_terms": ["disability", "WCB"]
  },
  
  // === SASKATCHEWAN ===
  "skca": {
    "name": "Saskatchewan Court of Appeal",
    "database": "skca",
    "jurisdiction": "SK",
    "search_terms": ["disability", "WCB"]
  },
  
  // === MANITOBA ===
  "mbca": {
    "name": "Manitoba Court of Appeal",
    "database": "mbca",
    "jurisdiction": "MB",
    "search_terms": ["disability", "WCB", "accommodation"]
  },
  
  // === QUEBEC ===
  "qctat": {
    "name": "Tribunal administratif du travail (Quebec)",
    "database": "qctat",
    "jurisdiction": "QC",
    "search_terms": ["disability", "CNESST", "accommodation"]
  },
  "qcca": {
    "name": "Quebec Court of Appeal",
    "database": "qcca",
    "jurisdiction": "QC",
    "search_terms": ["disability"]
  },
  
  // === NEW BRUNSWICK ===
  "nbca": {
    "name": "New Brunswick Court of Appeal",
    "database": "nbca",
    "jurisdiction": "NB",
    "search_terms": ["disability", "WorkSafeNB"]
  },
  
  // === NOVA SCOTIA ===
  "nsca": {
    "name": "Nova Scotia Court of Appeal",
    "database": "nsca",
    "jurisdiction": "NS",
    "search_terms": ["disability", "WCB"]
  },
  
  // === PRINCE EDWARD ISLAND ===
  "peca": {
    "name": "Prince Edward Island Court of Appeal",
    "database": "peca",
    "jurisdiction": "PE",
    "search_terms": ["disability", "WCB"]
  },
  
  // === NEWFOUNDLAND AND LABRADOR ===
  "nlca": {
    "name": "Newfoundland and Labrador Court of Appeal",
    "database": "nlca",
    "jurisdiction": "NL",
    "search_terms": ["disability", "WorkplaceNL"]
  },
  
  // === YUKON ===
  "ykca": {
    "name": "Yukon Court of Appeal",
    "database": "ykca",
    "jurisdiction": "YT",
    "search_terms": ["disability"]
  },
  
  // === NORTHWEST TERRITORIES ===
  "nwtca": {
    "name": "Northwest Territories Court of Appeal",
    "database": "nwtca",
    "jurisdiction": "NT",
    "search_terms": ["disability"]
  },
  
  // === NUNAVUT ===
  "nuca": {
    "name": "Nunavut Court of Appeal",
    "database": "nuca",
    "jurisdiction": "NU",
    "search_terms": ["disability"]
  },
  
  // === FEDERAL ===
  "chrt": {
    "name": "Canadian Human Rights Tribunal",
    "database": "chrt",
    "jurisdiction": "FED",
    "search_terms": ["accommodation", "disability"]
  },
  "fct": {
    "name": "Federal Court of Canada",
    "database": "fct",
    "jurisdiction": "FED",
    "search_terms": ["disability", "Canada Pension Plan"]
  },
  "fca": {
    "name": "Federal Court of Appeal",
    "database": "fca",
    "jurisdiction": "FED",
    "search_terms": ["disability", "CPP"]
  }
};

// ===== HELPER FUNCTIONS =====

function extractCondition(text) {
  const conditions = [
    "fibromyalgia", "chronic pain", "PTSD", "post-traumatic stress",
    "back injury", "spinal injury", "herniated disc", "depression",
    "anxiety", "rheumatoid arthritis", "osteoarthritis", "chronic fatigue",
    "multiple sclerosis", "MS", "carpal tunnel", "tendinitis"
  ];
  
  const textLower = text.toLowerCase();
  const found = conditions.filter(c => textLower.includes(c.toLowerCase()));
  
  return found.length > 0 ? found.join(", ") : "Unknown";
}

function extractOutcome(text) {
  const textLower = text.toLowerCase();
  
  if (/\bappeal.*allowed\b|\ballowed\b/.test(textLower)) return "Allowed";
  if (/\bappeal.*dismissed\b|\bdismissed\b/.test(textLower)) return "Dismissed";
  if (/\bdenied\b/.test(textLower)) return "Denied";
  if (/\bvaried\b/.test(textLower)) return "Varied";
  if (/\bremanded\b/.test(textLower)) return "Remanded";
  
  return "Unknown";
}

function extractEvidence(text) {
  const evidence = [];
  const textLower = text.toLowerCase();
  
  const evidenceTypes = [
    { pattern: /\brfc\b|residual functional capacity/i, name: "RFC form" },
    { pattern: /\bfce\b|functional capacity evaluation/i, name: "FCE" },
    { pattern: /specialist report|rheumatologist|psychiatrist|psychologist/i, name: "Specialist report" },
    { pattern: /timeline|chronology|history/i, name: "Timeline/history" },
    { pattern: /\bime\b|independent medical exam/i, name: "IME" },
    { pattern: /medical records|treatment history/i, name: "Medical records" },
    { pattern: /employer statement|job description/i, name: "Employment docs" }
  ];
  
  evidenceTypes.forEach(({ pattern, name }) => {
    if (pattern.test(text)) evidence.push(name);
  });
  
  return evidence;
}

function extractKeyFactors(text) {
  const factors = [];
  const textLower = text.toLowerCase();
  
  const factorPatterns = [
    { pattern: /credible|credibility established/i, name: "Credible testimony", success: true },
    { pattern: /consistent with medical evidence/i, name: "Consistent evidence", success: true },
    { pattern: /objective medical findings/i, name: "Objective findings", success: true },
    { pattern: /pre-existing condition/i, name: "Pre-existing condition noted", success: false },
    { pattern: /insufficient evidence/i, name: "Insufficient evidence", success: false },
    { pattern: /not credible|credibility concerns/i, name: "Credibility issues", success: false }
  ];
  
  factorPatterns.forEach(({ pattern, name, success }) => {
    if (pattern.test(text)) factors.push({ factor: name, success_indicator: success });
  });
  
  return factors;
}

// ===== API FUNCTIONS =====

function httpsGet(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          resolve(data);
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    }).on('error', reject);
  });
}

async function searchCanLII(database, searchTerm, offset = 0, retries = 3) {
  const params = new URLSearchParams({
    api_key: CANLII_API_KEY,
    offset: offset.toString(),
    resultCount: BATCH_SIZE.toString(),
    changedSince: CHANGED_SINCE,
    search: searchTerm
  });
  
  const url = `${CANLII_BASE}/caseBrowse/en/${database}?${params}`;
  
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      const data = await httpsGet(url);
      const json = JSON.parse(data);
      // CanLII returns "cases" not "results"
      return { results: json.cases || [] };
    } catch (error) {
      if (error.message.includes('429') && attempt < retries) {
        const waitTime = Math.pow(2, attempt) * 5000; // 5s, 10s, 20s
        console.log(`  ⏳ Throttled. Waiting ${waitTime/1000}s before retry ${attempt + 1}/${retries}...`);
        await new Promise(resolve => setTimeout(resolve, waitTime));
      } else {
        console.error(`  ❌ Error searching CanLII: ${error.message}`);
        return { results: [] };
      }
    }
  }
  
  return { results: [] };
}

async function fetchDecisionHTML(caseId, database) {
  const url = `${CANLII_BASE}/caseBrowse/en/${database}/${caseId}?api_key=${CANLII_API_KEY}`;
  
  try {
    const data = await httpsGet(url);
    // Parse JSON response and extract the HTML field
    const jsonResponse = JSON.parse(data);
    return jsonResponse.html || data; // Return HTML field, or raw data as fallback
  } catch (error) {
    console.error(`  ❌ Error fetching decision: ${error.message}`);
    return null;
  }
}

function parseDecision(caseData, html, tribunalName) {
  const text = html.replace(/<[^>]*>/g, ' ').replace(/\s+/g, ' ');
  
  return {
    case_id: caseData.caseId?.en || "Unknown",
    title: caseData.title || "Unknown",
    date: caseData.decisionDate || "Unknown",
    tribunal: tribunalName,
    url: caseData.url || "",
    condition: extractCondition(text),
    outcome: extractOutcome(text),
    evidence_cited: extractEvidence(text),
    key_factors: extractKeyFactors(text),
    snippet: text.substring(0, 500).trim() + "...",
    extraction_version: "v4.0-fulltext",  // Track which scraper version collected this
    raw_html: html.substring(0, 2000)  // Save more context for validation/debugging
  };
}

async function scrapeTribunal(tribunalId, config, maxResults = 100000) {
  console.log(`\n📊 Scraping ${config.name}...`);
  const allDecisions = [];
  
  for (const searchTerm of config.search_terms) {
    console.log(`  🔍 Searching for: ${searchTerm}`);
    let offset = 0;
    
    while (allDecisions.length < maxResults) {
      const results = await searchCanLII(config.database, searchTerm, offset);
      const cases = results.results || [];
      
      if (cases.length === 0) {
        console.log(`    ℹ️  No more results for '${searchTerm}'`);
        break;
      }
      
      const limit = Math.min(50, maxResults - allDecisions.length);
      for (let i = 0; i < Math.min(cases.length, limit); i++) {
        const caseData = cases[i];
        const caseId = caseData. caseId?.en;
        
        if (!caseId) continue;
        
        console.log(`    📄 Fetching ${caseId}...`);
        const html = await fetchDecisionHTML(caseId, config.database);
        
        if (html) {
          const decision = parseDecision(caseData, html, config.name);
          allDecisions.push(decision);
          console.log(`      ✅ ${decision.outcome} - ${decision.condition}`);
        }
        
        // Rate limiting - be nice to CanLII (3 seconds to avoid quota)
        await new Promise(resolve => setTimeout(resolve, 3000));
      }
      
      offset += BATCH_SIZE;
      if (cases.length < BATCH_SIZE) break;
    }
  }
  
  console.log(`  ✅ Scraped ${allDecisions.length} decisions from ${config.name}`);
  return allDecisions;
}

function saveDecisions(decisions, filename) {
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
  
  const filepath = path.join(OUTPUT_DIR, filename);
  fs.writeFileSync(filepath, JSON.stringify(decisions, null, 2));
  console.log(`  💾 Saved to ${filepath}`);
}

function generateSummary(allResults) {
  const summary = {
    total_decisions: 0,
    by_tribunal: {},
    by_outcome: {},
    by_condition: {},
    generated_at: new Date().toISOString()
  };
  
  Object.entries(allResults).forEach(([tribunalId, decisions]) => {
    summary.total_decisions += decisions.length;
    summary.by_tribunal[tribunalId] = decisions.length;
    
    decisions.forEach(d => {
      summary.by_outcome[d.outcome] = (summary.by_outcome[d.outcome] || 0) + 1;
      summary.by_condition[d.condition] = (summary.by_condition[d.condition] || 0) + 1;
    });
  });
  
  return summary;
}

// ===== MAIN =====

async function main() {
  console.log("=".repeat(60));
  console.log("🇨🇦 CanLII Remaining Provinces Scraper");
  console.log("=".repeat(60));
  console.log(`Output directory: ${OUTPUT_DIR}`);
  console.log(`Max results per tribunal: UNLIMITED (up to 100,000)`);
  console.log(`Date range: ${CHANGED_SINCE} to today`);
  console.log(`Tribunals/Courts: ${Object.keys(TRIBUNALS).length} (Ontario already collected)`);
  console.log();
  console.log("⏭️  SKIPPING: Ontario (onwsiat, onhrt, onca) - Already have 4,632 decisions");
  console.log("🆕 COLLECTING: BC, AB, SK, MB, QC, Atlantic, Territories, Federal");
  console.log();
  
  if (CANLII_API_KEY === "YOUR_FREE_API_KEY_HERE") {
    console.log("⚠️  WARNING: Set CANLII_API_KEY environment variable");
    console.log("   Example: $env:CANLII_API_KEY = \"your-key-here\"");
    console.log();
    process.exit(1);
  }
  
  const allResults = {};
  
  // Scrape each tribunal
  for (const [tribunalId, config] of Object.entries(TRIBUNALS)) {
    const decisions = await scrapeTribunal(tribunalId, config, 100000);
    allResults[tribunalId] = decisions;
    
    // Save individual tribunal results
    const date = new Date().toISOString().split('T')[0].replace(/-/g, '');
    const filename = `${tribunalId}-historical-${date}.json`;
    saveDecisions(decisions, filename);
  }
  
  // Generate and save summary
  const summary = generateSummary(allResults);
  console.log("\n" + "=".repeat(60));
  console.log("📊 SUMMARY");
  console.log("=".repeat(60));
  console.log(`Total decisions scraped: ${summary.total_decisions}`);
  console.log(`\nBy tribunal:`);
  Object.entries(summary.by_tribunal).forEach(([id, count]) => {
    console.log(`  ${id}: ${count} decisions`);
  });
  console.log(`\nBy outcome:`);
  Object.entries(summary.by_outcome).forEach(([outcome, count]) => {
    console.log(`  ${outcome}: ${count} (${(count/summary.total_decisions*100).toFixed(1)}%)`);
  });
  
  const date = new Date().toISOString().split('T')[0].replace(/-/g, '');
  saveDecisions(summary, `summary-historical-${date}.json`);
  
  console.log("\n✅ Scraping complete!");
  console.log(`📁 Results saved to: ${OUTPUT_DIR}`);
  console.log("\n🚀 Next: Analyze patterns with pattern-detection scripts");
}

main().catch(error => {
  console.error("\n❌ Fatal error:", error);
  process.exit(1);
});

File: _deploy/scripts/scrape-canlii-tribunals.py
#!/usr/bin/env python3
"""
CanLII Tribunal Decision Scraper
Scrapes WSIAT, HRTO, SST decisions to seed the 3 Flywheels

FREE TOOLS ONLY - NO API COSTS
- CanLII API: Free
- BeautifulSoup: Free
- Regex extraction: Free (no GPT-4 needed)

Author: 3mpwrApp
Date: April 2026
"""

import requests
import json
import re
import time
from datetime import datetime
from typing import List, Dict, Optional
import os

# ===== CONFIGURATION =====
CANLII_BASE = "https://api.canlii.org/v1"
CANLII_API_KEY = os.getenv("CANLII_API_KEY", "YOUR_FREE_API_KEY_HERE")  # Register at canlii.org/en/info/api.html
OUTPUT_DIR = "data/tribunal-decisions"
BATCH_SIZE = 50  # Be respectful to CanLII's servers

# Tribunals to scrape
TRIBUNALS = {
    "onwsiat": {
        "name": "Workplace Safety & Insurance Appeals Tribunal (Ontario)",
        "database": "onwsiat",
        "jurisdiction": "ON",
        "search_terms": ["fibromyalgia", "chronic pain", "PTSD", "back injury", "MSK"]
    },
    "sst": {
        "name": "Social Security Tribunal (CPP-D)",
        "database": "sst",
        "jurisdiction": "CA",
        "search_terms": ["disability", "CPP", "unable to work"]
    },
    "onhrt": {
        "name": "Human Rights Tribunal of Ontario",
        "database": "onhrt",
        "jurisdiction": "ON",
        "search_terms": ["accommodation", "disability discrimination"]
    }
}


# ===== HELPER FUNCTIONS =====

def extract_condition(text: str) -> Optional[str]:
    """Extract medical condition from decision text using regex patterns"""
    conditions = [
        "fibromyalgia", "chronic pain", "PTSD", "post-traumatic stress",
        "back injury", "spinal injury", "herniated disc", "depression",
        "anxiety", "rheumatoid arthritis", "osteoarthritis", "chronic fatigue",
        "multiple sclerosis", "MS", "carpal tunnel", "tendinitis"
    ]
    
    text_lower = text.lower()
    found = []
    
    for condition in conditions:
        if condition.lower() in text_lower:
            found.append(condition)
    
    return ", ".join(found) if found else "Unknown"


def extract_outcome(text: str) -> str:
    """Extract decision outcome from text"""
    text_lower = text.lower()
    
    # Priority order (most specific first)
    if re.search(r'\bappeal.*allowed\b|\ballowed\b', text_lower):
        return "Allowed"
    elif re.search(r'\bappeal.*dismissed\b|\bdismissed\b', text_lower):
        return "Dismissed"
    elif re.search(r'\bappeal.*denied\b|\bdenied\b', text_lower):
        return "Denied"
    elif re.search(r'\bvaried\b', text_lower):
        return "Varied"
    elif re.search(r'\bremanded\b', text_lower):
        return "Remanded"
    else:
        return "Unknown"


def extract_evidence(text: str) -> List[str]:
    """Extract types of evidence mentioned"""
    evidence_patterns = {
        "RFC form": r"\bRFC\b|\bresidual functional capacity\b",
        "Functional capacity evaluation": r"\bfunctional capacity evaluation\b|\bFCE\b",
        "Specialist report": r"\bspecialist report\b|\bconsultant.*report\b",
        "Timeline": r"\btimeline\b|\bchronology\b",
        "Medical records": r"\bmedical records?\b|\bclinical notes\b",
        "Vocational assessment": r"\bvocational assessment\b|\bwork capacity\b",
        "IME report": r"\bIME\b|\bindependent medical exam",
        "Employer statement": r"\bemployer.*statement\b|\bemployer.*report\b"
    }
    
    found = []
    text_lower = text.lower()
    
    for evidence_type, pattern in evidence_patterns.items():
        if re.search(pattern, text_lower):
            found.append(evidence_type)
    
    return found


def extract_key_factors(text: str) -> List[str]:
    """Extract success/failure factors from decision"""
    factors = []
    text_lower = text.lower()
    
    # Success factors
    if "treating physician" in text_lower:
        factors.append("Treating physician testimony was persuasive")
    if "credible" in text_lower and "witness" in text_lower:
        factors.append("Witness credibility established")
    if "consistent" in text_lower:
        factors.append("Evidence was internally consistent")
    
    # Failure factors
    if "pre-existing" in text_lower:
        factors.append("Pre-existing condition argument raised")
    if "insufficient" in text_lower and "evidence" in text_lower:
        factors.append("Evidence deemed insufficient")
    if "non-compliant" in text_lower:
        factors.append("Worker deemed non-compliant")
    
    return factors


# ===== MAIN SCRAPING FUNCTIONS =====

def search_canlii(database: str, search_term: str, offset: int = 0) -> Dict:
    """Search CanLII API for decisions"""
    url = f"{CANLII_BASE}/caseBrowse/en/{database}"
    params = {
        "api_key": CANLII_API_KEY,
        "offset": offset,
        "resultCount": BATCH_SIZE,
        "search": search_term
    }
    
    try:
        response = requests.get(url, params=params, timeout=30)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f"  ❌ Error searching CanLII: {e}")
        return {"results": []}


def fetch_decision_html(case_id: str, database: str) -> Optional[str]:
    """Fetch full HTML of a decision"""
    url = f"{CANLII_BASE}/caseBrowse/en/{database}/{case_id}"
    params = {"api_key": CANLII_API_KEY}
    
    try:
        response = requests.get(url, params=params, timeout=30)
        response.raise_for_status()
        data = response.json()
        return data.get("html", "")
    except Exception as e:
        print(f"    ❌ Error fetching case {case_id}: {e}")
        return None


def parse_decision(case_data: Dict, html: str, tribunal: str) -> Dict:
    """Parse decision into structured format"""
    return {
        "id": case_data.get("caseId", {}).get("en", "unknown"),
        "title": case_data.get("title", "Untitled"),
        "date": case_data.get("decisionDate", ""),
        "tribunal": tribunal,
        "condition": extract_condition(html),
        "outcome": extract_outcome(html),
        "evidence_cited": extract_evidence(html),
        "key_factors": extract_key_factors(html),
        "url": case_data.get("url", ""),
        "scraped_at": datetime.now().isoformat()
    }


def scrape_tribunal(tribunal_id: str, config: Dict, max_results: int = 100) -> List[Dict]:
    """Scrape decisions from a specific tribunal"""
    print(f"\n📊 Scraping {config['name']}...")
    all_decisions = []
    
    for search_term in config['search_terms']:
        print(f"  🔍 Searching for: {search_term}")
        offset = 0
        
        while len(all_decisions) < max_results:
            # Search
            results = search_canlii(config['database'], search_term, offset)
            cases = results.get("results", [])
            
            if not cases:
                print(f"    ℹ️ No more results for '{search_term}'")
                break
            
            # Process each case
            for case in cases[:min(10, max_results - len(all_decisions))]:
                case_id = case.get("caseId", {}).get("en")
                if not case_id:
                    continue
                
                print(f"    📄 Fetching {case_id}...")
                html = fetch_decision_html(case_id, config['database'])
                
                if html:
                    decision = parse_decision(case, html, config['name'])
                    all_decisions.append(decision)
                    print(f"      ✅ {decision['outcome']} - {decision['condition']}")
                
                # Rate limiting
                time.sleep(1)  # Be nice to CanLII servers
            
            offset += BATCH_SIZE
            
            if len(cases) < BATCH_SIZE:
                break
    
    print(f"  ✅ Scraped {len(all_decisions)} decisions from {config['name']}")
    return all_decisions


def save_decisions(decisions: List[Dict], filename: str):
    """Save decisions to JSON file"""
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    filepath = os.path.join(OUTPUT_DIR, filename)
    
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(decisions, f, indent=2, ensure_ascii=False)
    
    print(f"  💾 Saved to {filepath}")


def generate_summary(decisions: List[Dict]) -> Dict:
    """Generate summary statistics"""
    total = len(decisions)
    
    outcomes = {}
    conditions = {}
    evidence_types = {}
    
    for decision in decisions:
        # Count outcomes
        outcome = decision['outcome']
        outcomes[outcome] = outcomes.get(outcome, 0) + 1
        
        # Count conditions
        condition = decision['condition']
        conditions[condition] = conditions.get(condition, 0) + 1
        
        # Count evidence types
        for evidence in decision['evidence_cited']:
            evidence_types[evidence] = evidence_types.get(evidence, 0) + 1
    
    # Calculate success rate (Allowed / Total)
    allowed = outcomes.get("Allowed", 0)
    success_rate = (allowed / total * 100) if total > 0 else 0
    
    return {
        "total_decisions": total,
        "success_rate": f"{success_rate:.1f}%",
        "outcomes": outcomes,
        "conditions": conditions,
        "evidence_types": evidence_types,
        "most_common_condition": max(conditions.items(), key=lambda x: x[1])[0] if conditions else "None",
        "most_cited_evidence": max(evidence_types.items(), key=lambda x: x[1])[0] if evidence_types else "None"
    }


# ===== MAIN EXECUTION =====

def main():
    """Main execution"""
    print("=" * 60)
    print("🔄 CanLII Tribunal Decision Scraper")
    print("=" * 60)
    print(f"Output directory: {OUTPUT_DIR}")
    print(f"Max results per tribunal: 100")
    print()
    
    if CANLII_API_KEY == "YOUR_FREE_API_KEY_HERE":
        print("⚠️  WARNING: Set CANLII_API_KEY environment variable")
        print("   Register for free at: https://www.canlii.org/en/info/api.html")
        print()
        # Continue anyway for demo purposes
    
    all_results = {}
    
    # Scrape each tribunal
    for tribunal_id, config in TRIBUNALS.items():
        decisions = scrape_tribunal(tribunal_id, config, max_results=100)
        all_results[tribunal_id] = decisions
        
        # Save individual tribunal results
        filename = f"{tribunal_id}-decisions-{datetime.now().strftime('%Y%m%d')}.json"
        save_decisions(decisions, filename)
        
        # Generate summary
        summary = generate_summary(decisions)
        print(f"\n  📈 Summary for {config['name']}:")
        print(f"     Total: {summary['total_decisions']}")
        print(f"     Success rate: {summary['success_rate']}")
        print(f"     Most common condition: {summary['most_common_condition']}")
        print(f"     Most cited evidence: {summary['most_cited_evidence']}")
    
    # Combine all results
    combined = []
    for decisions in all_results.values():
        combined.extend(decisions)
    
    save_decisions(combined, f"all-tribunals-{datetime.now().strftime('%Y%m%d')}.json")
    
    # Final summary
    print("\n" + "=" * 60)
    print("✅ SCRAPING COMPLETE")
    print("=" * 60)
    print(f"Total decisions scraped: {len(combined)}")
    print(f"Output directory: {OUTPUT_DIR}")
    print()
    print("Next steps:")
    print("1. Review data/tribunal-decisions/*.json")
    print("2. Import to database")
    print("3. Run pattern detection")
    print("4. Generate templates from successful cases")


if __name__ == "__main__":
    main()
